core/src/main/kotlin/knote/PageManagerImpl.kt

```diff
@@ -0,0 +1,314 @@
+package knote
+
+import knote.api.Page
+import knote.api.PageManager
+import knote.data.NotebookImpl
+import knote.data.PageImpl
+import knote.host.EvalScript
+import knote.script.PageScript
+import knote.util.MutableKObservableMap
+import knote.util.watchActor
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import mu.KLogging
+import java.io.File
+import java.lang.Exception
+import kotlin.reflect.full.declaredFunctions
+import kotlin.script.experimental.api.ScriptDiagnostic
+import kotlin.script.experimental.jvmhost.BasicJvmScriptingHost
+
+internal class PageManagerImpl(
+    private val notebook: NotebookImpl,
+    private val host: BasicJvmScriptingHost,
+    private val workingDir: File
+) : PageManager {
+    companion object : KLogging()
+
+    val notebookScript get() = notebook.compiledScript!!
+
+    override val pages: MutableKObservableMap<String, Page> = MutableKObservableMap()
+    override val compiledPages: MutableKObservableMap<String, PageScript> = MutableKObservableMap()
+    override val results: MutableKObservableMap<String, Any> = MutableKObservableMap()
+
+    override val dependencies: MutableKObservableMap<String, Set<String>> = MutableKObservableMap()
+    override val reportMap: MutableKObservableMap<String, List<ScriptDiagnostic>> = MutableKObservableMap()
+    override val fileContentMap: MutableKObservableMap<String, String> = MutableKObservableMap()
+
+    init {
+
+        startWatcher()
+
+        // loop over all pages that are evaluated but have no result yet
+
+    }
+
+    override fun executeAll() : Map<String, Any>{
+        notebookScript.pageFiles.forEach {
+            evalPage(it)
+        }
+        while (pages.any { (id, page) -> page.result == null && !(page as PageImpl).errored}) {
```

Comment:
`result == null` is just for filtering pages that do not have a result yet.. so they need to execute or re-execute
it also checks if it errored otherwise it will force into a infinite loop when some pages do not compile

it seems hte forEach call there used to be a `assosciate` call